8.0 refactor picking invoice wizard

leorochael opened the pull request:
Descrição do problema/nova funcionalidade deste Pull Resquest(PR):
------------------------------------------------------------------

Este PR simplifica a maneira como o wizard de criação de faturas a partir de pickings localiza os documentos relacionados.

Este PR visa preparar o caminho para uma modificação deste wizard que permita uma única fatura relacionar vários documentos (outras faturas ou pos.orders).

Comportamento atual antes do PR:
--------------------------------

Os invoices e pos.orders relacionados eram buscados via chave de acesso e o código era bem mais rebuscado

Comportamento esperado depois do PR:
------------------------------------

Os invoices e pos.orders são buscado diretamente por relacionamentos, em código que é bem mais simples de compreender.

- [x] Esta mudança não altera a estrutura do banco de dados, portanto não precisa de script de migração.

--
Eu confirmo que eu assinei a CLA e li as recomendações de como contribuir:
- https://odoo-community.org/page/cla
- https://odoo-community.org/page/Contribute

leorochael committed:
[REF] Picking Invoice Wizard: delegação

Refatora delegação de métodos do Wizard `stock.invoice.onshipping` para
evitar repetição do campo em subclasses.

leorochael committed:
[REF] Simplifica cálculo do fiscal_doc_ref default

No wizard `stock.invoice.onshipping` que cria faturas para pickings,
simplifica código que calcula os documentos relacionados, sem alterar
a semântica.

Para de depender da busca do fiscal_document_access_key e passa a
depender diretamente dos relacionamentos com os documentos originais.

leorochael commented:
Favor analizar este PR commit a commit, ao invés de todas as mudanças ao mesmo tempo.

mileo commented on l10n_br_stock_account/wizard/stock_invoice_onshipping.py:
```diff
@@ -16,28 +16,49 @@ class StockInvoiceOnShipping(models.TransientModel):
+    @api.multi
+    def _fiscal_doc_ref_default(self):
+        return_picking_ids = self.get_returned_picking_ids()
+        ref_id = return_picking_ids.mapped('invoice_ids')[:1].id
```
Não seria melhor:
ref_id = return_picking_ids.mapped('invoice_ids').ids[:1] ou apenas .id ?

leorochael replied:
> Não seria melhor:
>     `ref_id = return_picking_ids.mapped('invoice_ids').ids[:1]`

Não, pois `.ids` é uma lista que, nesse caso, poderia ser vazia, ter um elemento, ou ter vários elementos, ao passo que `recordset[:1]` é sempre um recordset, que pode ser vazio, ou ter um elemento.

Se `recordset[:1]` for vazio, `ref_id = recordset[:1].id` vai ser `False` o que significa que a interpolação:

    res = 'account.invoice,%d' % ref_id

Terá como resultado:

    account.invoice,0

Que é exatamente o que queremos nesse caso. Por outro lado, se pegássemos `.ids[:1]` teríamos uma lista que pode estar vazia, ou conter um elemento, e seria necessário um `if` (ou no mínimo um `or`) a mais para distinguir os dois casos e pegar o id correto.

> ou apenas `.id`?

Se o recordset resultante de `return_picking_ids.mapped('invoice_ids')` tiver múltiplos elementos, então `return_picking_ids.mapped('invoice_ids').id` vai levantar uma exceção:

    except_orm: ('ValueError', 'Expected singleton: account.invoice(x, y, z,...)')

Ao passo que `return_picking_ids.mapped('invoice_ids')[0].id` vai levantar outra exceção se `return_picking_ids.mapped('invoice_ids')` estiver vazio:

    IndexError: tuple index out of range

A solução atual é a única que não envolve nenhuma condicional.

leorochael committed:
[IMP] Faturas: múltiplos documentos relacionados

No wizard de criação de faturas, permite relacionar mais de
um documento de cada vez.

Subclasses que queiram relacionar mais de um tipo de documento agora
só precisam listar qual o relacionamento com o picking original e como
mapear os dados do documento relacionado.
